src/client/dfuse/pil4dfs/dfs_dcache.c

```diff
@@ -0,0 +1,663 @@
+/**
+ * (C) Copyright 2022-2024 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#define D_LOGFAC DD_FAC(il)
+
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <daos/debug.h>
+#include <gurt/hash.h>
+#include <gurt/list.h>
+#include <daos_fs.h>
+#include <daos/common.h>
+
+#include "dfs_dcache.h"
+
+/** Size of the hash key prefix */
+#define DCACHE_KEY_PREF_SIZE 35
+#define DCACHE_KEY_MAX       (DCACHE_KEY_PREF_SIZE - 1 + PATH_MAX)
+
+#ifdef DAOS_BUILD_RELEASE
+
+#define DF_DK          "dk[%zi]"
+#define DP_DK(_dk)     strnlen((_dk), DCACHE_KEY_MAX)
+
+#define DF_PATH        "path[%zi]"
+#define DP_PATH(_path) strnlen((_path), PATH_MAX)
+
+#else
+
+#define DF_DK          "dk'%s'"
+#define DP_DK(_dk)     (_dk)
+
+#define DF_PATH        "path'%s'"
+#define DP_PATH(_path) (_path)
+
+#endif
+
+typedef int (*destroy_fn_t)(dfs_dcache_t *);
+typedef int (*find_insert_fn_t)(dfs_dcache_t *, char *, size_t, dcache_rec_t **);
+typedef void (*drec_incref_fn_t)(dfs_dcache_t *, dcache_rec_t *);
+typedef void (*drec_decref_fn_t)(dfs_dcache_t *, dcache_rec_t *);
+typedef void (*drec_del_at_fn_t)(dfs_dcache_t *, dcache_rec_t *);
+typedef int (*drec_del_fn_t)(dfs_dcache_t *, char *, dcache_rec_t *);
+
+/** DFS directory cache */
+struct dfs_dcache {
+	/** Cached DAOS file system */
+	dfs_t              *dd_dfs;
+	/** Hash table holding the cached directories */
+	struct d_hash_table dd_dir_hash;
+	/** Key prefix of the DFS root directory */
+	char                dd_key_root_prefix[DCACHE_KEY_PREF_SIZE];
+	/** Destroy a dfs dir-cache */
+	destroy_fn_t        destroy_fn;
+	/** Return the dir-cahe record of a given location and insert it if needed */
+	find_insert_fn_t    find_insert_fn;
+	/** Increase the reference counter of a given dir-cache record */
+	drec_incref_fn_t    drec_incref_fn;
+	/** Decrease the reference counter of a given dir-cache record */
+	drec_decref_fn_t    drec_decref_fn;
+	/** Delete a given dir-cache record */
+	drec_del_at_fn_t    drec_del_at_fn;
+	/** Delete the dir-cache record of a given location */
+	drec_del_fn_t       drec_del_fn;
+};
+
+/** Entry of a DFS directory cache */
+struct dcache_rec {
+	/** Entry in the hash table of the DFS cache */
+	d_list_t         dr_entry;
```

Comment:
Is there a requirement for this entry to be first in the struct?  If so it should be commented as such.

Reply:
There is no requirements from what I know.
No call to containerof or such struct pointer trick until now.